pkg/component/component.go

```diff
@@ -0,0 +1,66 @@
+package component
+
+import (
+	"strings"
+
+	"github.com/improbable-eng/thanos/pkg/store/storepb"
+)
+
+type StoreAPI interface {
+	implementsStoreAPI()
+	String() string
+	ToProto() storepb.Component
+}
+
+type Source interface {
+	producesBlocks()
+	String() string
+}
+
+type SourceStoreAPI interface {
+	implementsStoreAPI()
+	producesBlocks()
+	String() string
+	ToProto() storepb.Component
+}
+
+type component struct {
+	name string
+}
+
+func (c component) String() string { return c.name }
+
+type storeAPI struct {
+	component
+}
+
+func (storeAPI) implementsStoreAPI() {}
+
+func (s sourceStoreAPI) ToProto() storepb.Component {
+	return storepb.Component(storepb.Component_value[strings.ToUpper(s.String())])
+}
+
+func (s storeAPI) ToProto() storepb.Component {
+	return storepb.Component(storepb.Component_value[strings.ToUpper(s.String())])
+}
+
+type source struct {
+	component
+}
+
+func (source) producesBlocks() {}
+
+type sourceStoreAPI struct {
+	component
+	source
+	storeAPI
+}
+
+var (
+	Compact    = source{component: component{name: "compact"}}
+	Downsample = source{component: component{name: "downsample"}}
+	Query      = sourceStoreAPI{component: component{name: "query"}}
+	Rule       = sourceStoreAPI{component: component{name: "rule"}}
+	Sidecar    = sourceStoreAPI{component: component{name: "sidecar"}}
+	Store      = sourceStoreAPI{component: component{name: "store"}}
```

Comment:
Bucket command is a source technically. And commponent as well (: each subcommand of bucket cmd technically. (:

Reply:
Added the bucket source 😄

Reply:
We have different `bucket` sources though! (: `bucket.repair` and `bucket.sync` now. Can we have it as separate sources?